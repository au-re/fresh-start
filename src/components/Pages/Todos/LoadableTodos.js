// split the bundle on a page level to limit the size of the initial load
// https://github.com/jamiebuilds/react-loadable

import React from "react";
import Loadable from "react-loadable";

import Loading from "../Loading/Loading";

export default Loadable({
  loader: () => import("./Todos"),
  loading: () => <Loading />,
});

