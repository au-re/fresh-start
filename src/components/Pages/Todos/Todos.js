import { message } from "antd";
import React, { Component } from "react";
import { DragDropContext, Draggable, Droppable } from "react-beautiful-dnd";
import styled from "styled-components";
import request from "superagent";

import { fadeColor, fadeIn } from "../../../styles/animations";
import CardList from "../../CardList/CardList";

const { Card } = CardList;

const AnimatedCardList = styled(CardList) `
  opacity: 0;
  animation: ${fadeIn()} 0.4s ease-out forwards;
  animation-delay: 0.5s;
  background: ${(props) => (props.isDraggingOver ? "papayawhip" : "transparent")};
`;

const reorder = (list, startIndex, endIndex) => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);
  result.splice(endIndex, 0, removed);
  return result;
};

class Todos extends Component {

  state = {
    todos: [],
  }

  async componentDidMount() {
    try {
      const { body } = await request
        .get("https://jsonplaceholder.typicode.com/posts")
        .accept("application/json");
      this.setState({ todos: body });
    } catch (error) {
      message.error("something went wrong");
    }
  }

  onDragEnd = (result) => {
    // dropped outside the list
    if (!result.destination) return;

    const { todos } = this.state;
    const reorderedTodos = reorder(
      todos,
      result.source.index,
      result.destination.index,
    );

    this.setState({ todos: reorderedTodos });
  }

  render() {
    const { className } = this.props;
    const { todos } = this.state;
    return (
      <div className={className}>
        <DragDropContext onDragEnd={this.onDragEnd}>
          <Droppable droppableId="droppable">
            {
              (provided, snapshot) => (
                <div ref={provided.innerRef}>
                  <AnimatedCardList isDraggingOver={snapshot.isDraggingOver}>
                    {
                      todos.map((todo, idx) =>
                        <Draggable key={todo.id} draggableId={todo.id} index={idx}>
                          {
                            (provided, snapshot) => (
                              <div
                                ref={provided.innerRef}
                                {...provided.draggableProps}
                                {...provided.dragHandleProps}>
                                <Card
                                  isDragging={snapshot.isDragging}
                                  title={todo.title}>
                                  {todo.body}
                                </Card>
                              </div>)
                          }
                        </Draggable>)
                    }
                    {provided.placeholder}
                  </AnimatedCardList>
                </div>)
            }
          </Droppable>
        </DragDropContext>

      </div>);
  }
}

export default styled(Todos) `
  min-height: 100%;
  background: coral;
  animation: ${fadeColor("coral", "sandybrown")} 1s ease-out forwards;
`;
