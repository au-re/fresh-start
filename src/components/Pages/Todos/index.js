import { message } from "antd";
import PropTypes from "prop-types";
import React, { Component } from "react";

class Todos extends Component {

  async componentDidMount() {
    const { fetchTodoById } = this.props;
    try {
      await fetchTodoById(1);
    } catch (error) {
      message.error("something went wrong");
    }
  }

  render() {
    const { className } = this.props;
    return (
      <div className={className}>
        todos
      </div>);
  }
}

Todos.propTypes = {
  fetchTodoById: PropTypes.func,
};

Todos.defaultProps = {
  fetchTodoById: () => { },
};

export default Todos;
