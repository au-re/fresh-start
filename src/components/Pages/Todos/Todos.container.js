import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { getResources, getStatus } from "redux-resource";

import { fetchTodoById } from "../../../redux/resources/todos";
import Todos from "../Todos/Todos";

const mapStateToProps = (state) => ({
  todos: getResources(state.todos, () => true),
  status: getStatus(state, "todos.requests.fetchTodos.status"),
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators({ fetchTodoById }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Todos);
