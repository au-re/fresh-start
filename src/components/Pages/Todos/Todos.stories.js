import { withInfo } from "@storybook/addon-info";
import { storiesOf } from "@storybook/react";
import React from "react";

import Todos from "./Todos";

storiesOf("Todos", module)

  .add("basic usage", withInfo(`Page that displays a list of todos`)(
    () => <Todos />));
