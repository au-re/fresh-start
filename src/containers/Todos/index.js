import { connect } from "react-redux";
import { bindActionCreators } from "redux";
import { getResources, getStatus } from "redux-resource";

import Todos from "../../components/Pages/Todos";
import { fetchTodoById } from "../../redux/resources";

const mapStateToProps = (state) => ({
  todos: getResources(state.todos, () => true),
  status: getStatus(state, "todos.requests.fetchTodos.status"),
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators({ fetchTodoById }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Todos);
