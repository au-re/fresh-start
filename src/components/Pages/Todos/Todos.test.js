import Enzyme from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import React from "react";

import Todos from "./Todos";

Enzyme.configure({ adapter: new Adapter() });

describe("Todos", () => {

  it("renders without props", () => {
    expect(<Todos />).toMatchSnapshot();
  });
});
