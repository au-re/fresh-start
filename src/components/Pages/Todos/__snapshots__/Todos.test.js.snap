// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos renders without props 1`] = `
<Todos
  fetchTodoById={[Function]}
/>
`;
